### PR TITLE
Add namespace exclusion for watchers

### PR DESCRIPTION
**Pull request by komapa**
The reflector watches all ConfigMaps and Secrets across every namespace in the cluster. In large clusters with thousands of ConfigMaps, this creates massive event volume — the vast majority of which are for resources with no reflector annotations. This unnecessary traffic saturates the internal bounded channel (256 slots) and can contribute to silent stalls.

This PR adds a new `excludedNamespaces` configuration option that accepts comma-separated glob patterns. Events from matching namespaces are dropped before entering the bounded channel, eliminating noise at the earliest possible point.

### CODE REVIEW

**Comment by komapa** [comment]
@winromulus would love to get this reviewed/merged if possible

**Review by winromulus — COMMENTED** [review]
Thanks for tackling this — the bounded-channel pressure in large clusters is a real operational pain point, and filtering at the producer side before the channel is the right layer to catch it.

A few observations from reading through, mostly small refinements plus one larger ask around test coverage.

## What works well

- Good choice of filter location — dropping events before `WriteAsync` means the channel never sees the noise.
- Logging is thoughtful: patterns at session start, excluded count at session close when non-zero. Useful operational signal.
- `IOptionsMonitor.CurrentValue` captured at session start matches how `WatcherTimeout` already behaves — consistent with the existing pattern.
- Helm wiring is complete across both `deployment.yaml` and `cron.yaml`, `values.yaml` has an explanatory comment, README table updated.
- `Regex.Escape` before replacing `\*` / `\?` is the correct glob-to-regex translation; `RegexOptions.Compiled` is appropriate for a per-event matcher.

## Main ask: tests

The feature currently has no coverage. Given the integration suite already uses `Testcontainers.K3s` with a clean builder pattern (see `MirroringIntegrationTests.AutoReflect_To_AllowedNamespaces`), this feels like a good candidate for:

- A unit-level test of the glob parser/matcher covering edge cases like empty strings, whitespace, `*` alone, multiple patterns, `?` wildcards, and namespace names containing regex metacharacters. `ParseGlobPatterns` and `IsNamespaceExcluded` are `private static` in the base class today — extracting them to a small `GlobMatcher` helper would make them trivially testable and slightly more reusable.
- An integration test mirroring the shape of `AutoReflect_To_AllowedNamespaces`: set `ES_Reflector__Watcher__ExcludedNamespaces=excluded-*`, create a source in an excluded namespace and verify no mirrors appear; create one in an allowed namespace and verify they do.

Without coverage, a future well-meaning simplification of the escape sequence could silently change `?` matching or similar.

## Scope clarification worth adding to the docs

One thing an operator might reasonably misread: this setting drops watch events *originating* in the listed namespaces, but it doesn't stop auto-reflection from creating mirrors *into* them — target-side filtering still runs off the source's `reflection-auto-namespaces` annotation. A sentence in the README covering that distinction would head off confusion.

## On `NamespaceWatcher`

Since `V1Namespace` is cluster-scoped, `item.Metadata.NamespaceProperty` is null for namespace events and the filter short-circuits. That's actually correct — you want `NamespaceWatcher` to keep seeing events so auto-reflection fires on new namespace creation — just flagging it so the interaction is explicit in the review record.

Smaller notes left inline. Overall a focused, well-scoped change solving a concrete problem.

---

Thanks again for the contribution — this is clearly solving a real-world pain point, and I appreciate the care taken with the Helm wiring and observability. Happy to merge once the items above are addressed.

File: README.md [hunk 1]
```diff
@@ -46,6 +46,7 @@ You can customize the values of the helm deployment by using the following Value
 | `image.pullPolicy`                       | Container image pull policy                      | `IfNotPresent`                                                                                   |
 | `configuration.logging.minimumLevel`     | Logging minimum level                            | `Information`                                                                                    |
 | `configuration.watcher.timeout`          | Maximum watcher lifetime in seconds              | ``                                                                                               |
+| `configuration.watcher.excludedNamespaces` | Comma-separated list of namespace glob patterns to exclude from watching. Supports `*` (any characters) and `?` (single character). Example: `"ephie-*,kube-system,*-temp"` | ``                                                                                               |
```
Comment (winromulus) [inline README.md]: Small wording suggestion: "exclude from watching" reads as if the watch request itself is narrowed server-side. In practice the watch API still streams every namespace over the wire and events are filtered client-side before entering the channel. Something like "exclude from reflection processing" or "exclude from event handling" might be a touch more precise. Not blocking.

File: src/ES.Kubernetes.Reflector/Watchers/Core/WatcherBackgroundService.cs [hunk 2]
```diff
@@ -107,4 +126,25 @@ await handler.Handle(new WatcherClosed
...
+
+    /// <summary>
+    ///     Parses a comma-separated list of glob patterns into compiled Regex objects.
+    ///     Supports * (any characters) and ? (single character).
+    /// </summary>
+    private static Regex[] ParseGlobPatterns(string? patterns)
+    {
+        if (string.IsNullOrWhiteSpace(patterns)) return [];
+        return patterns.Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
+            .Where(p => !string.IsNullOrWhiteSpace(p))
```
Comment (winromulus) [inline src/ES.Kubernetes.Reflector/Watchers/Core/WatcherBackgroundService.cs]: `StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries` already drops any entries that are empty after trimming, so this `.Where` never removes anything. Safe to drop for a slightly tighter pipeline.

File: src/ES.Kubernetes.Reflector/Watchers/Core/WatcherBackgroundService.cs [hunk 3]
```diff
@@ -62,6 +70,12 @@ await watcherEventHandler.Handle(new WatcherEvent
                 {
                     await foreach (var (type, item) in watchList)
                     {
+                        if (IsNamespaceExcluded(item.Metadata?.NamespaceProperty, excludedNamespacePatterns))
```
Comment (winromulus) [inline src/ES.Kubernetes.Reflector/Watchers/Core/WatcherBackgroundService.cs]: Worth a short code comment here: this filter is effectively a no-op for `NamespaceWatcher` since `V1Namespace.Metadata.NamespaceProperty` is always null for cluster-scoped resources (the `string.IsNullOrEmpty(ns)` check short-circuits). That's the desired behavior — namespace events should keep flowing so auto-reflection triggers on new namespace creation — but making the *why* self-documenting would help the next reader.

File: src/helm/reflector/values.yaml [hunk 4]
```diff
@@ -28,6 +28,10 @@ configuration:
     minimumLevel: Information
   watcher:
     timeout: ""
+    # Comma-separated list of namespace glob patterns to exclude from watching.
+    # Supports wildcards: * (any characters), ? (single character).
+    # Example: "kube-system,*-suffix,prefix-*"
+    excludedNamespaces: ""
```
Comment (winromulus) [inline src/helm/reflector/values.yaml:34]: Consider noting that pattern matching is case-sensitive. Kubernetes namespaces are DNS-1123 compliant (lowercase only) so in practice it's a non-issue, but an operator pasting `Kube-System` into config will get a confusing non-match. Alternative: add `RegexOptions.IgnoreCase` in `ParseGlobPatterns` to make it case-insensitive by default.

**Review by winromulus — CHANGES_REQUESTED** [review]

File: src/helm/reflector/values.yaml (hunk 4 above)
Comment (komapa) [reply src/helm/reflector/values.yaml:34]: Updated to lowercase the pattern before we convert it to a GlobPattern.

**Comment by komapa** [comment]
Thank you so much for the detailed review and directions. 

I have to admit, I used quite a lot of AI for the unit/integration tests so please take a closer look there:

```
❯ dotnet test tests/ES.Kubernetes.Reflector.Tests/ES.Kubernetes.Reflector.Tests.csproj --no-build -verbosity:minimal 2>&1
[xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v3.1.5+1b188a7b0a (64-bit .NET 10.0.5)
[xUnit.net 00:00:00.14]   Discovering: ES.Kubernetes.Reflector.Tests
[xUnit.net 00:00:00.25]   Discovered:  ES.Kubernetes.Reflector.Tests
[xUnit.net 00:00:00.36]   Starting:    ES.Kubernetes.Reflector.Tests
[xUnit.net 00:00:28.48]   Finished:    ES.Kubernetes.Reflector.Tests (ID = 'e0bea2b0f8b2c51a4d44fa5fcb5344e51fd55832d3ed1872674304864c6cfc26')
  ES.Kubernetes.Reflector.Tests test net10.0 succeeded (28.9s)

Test summary: total: 27, failed: 0, succeeded: 27, skipped: 0, duration: 28.9s
Build succeeded in 29.0s
```

File: src/ES.Kubernetes.Reflector/Watchers/Core/WatcherBackgroundService.cs [hunk 5]
```diff
@@ -36,9 +36,20 @@ protected override async Task ExecuteAsync(CancellationToken stoppingToken)
                 FullMode = BoundedChannelFullMode.Wait
             });
 
+            // Kubernetes namespace names must be valid DNS-1123 labels, which are lowercase-only,
+            // so normalizing the configured exclusion patterns to lowercase ensures comparisons
+            // against Metadata.NamespaceProperty are consistent without changing semantics.
+            var excludedNamespacePatterns = GlobMatcher.ParseGlobPatterns(options.CurrentValue.Watcher?.ExcludedNamespaces?.ToLower());
```
Comment (winromulus) [inline src/ES.Kubernetes.Reflector/Watchers/Core/WatcherBackgroundService.cs]: Tiny nit: `ToLower()` is culture-sensitive — in `tr-TR` locale, `"I".ToLower()` returns the dotless i (`ı`) rather than `i`, which would produce a regex that doesn't match what an operator expected. Since namespace names are ASCII DNS-1123, `ToLowerInvariant()` is the safer choice here. One-character fix, purely defensive.

File: tests/ES.Kubernetes.Reflector.Tests/Integration/ExcludedNamespacesIntegrationTests.cs [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        var excludedSourceResource = await CreateResource(client, namespaceName: excludedNamespace,
+            annotations: new ReflectorAnnotationsBuilder()
+                .WithReflectionAllowed(true)
+                .WithAutoEnabled(true).Build());
+
+        // Resource in allowed namespace — should mirror normally
+        var allowedSourceResource = await CreateResource(client, namespaceName: allowedNamespace,
+            annotations: new ReflectorAnnotationsBuilder()
+                .WithReflectionAllowed(true)
+                .WithAllowedNamespaces($"^{allowedNamespace}$")
```
Comment (winromulus) [inline tests/ES.Kubernetes.Reflector.Tests/Integration/ExcludedNamespacesIntegrationTests.cs]: Small thing on test strength: the positive-control source is scoped here to its own namespace via `^{allowedNamespace}$`, so it's not actually exercising cross-namespace auto-reflection. A regression that silently broke reflection entirely could still pass both assertions: the excluded source correctly doesn't propagate (negative ✓), and the allowed source correctly reflects within its own namespace (same-namespace, happens with or without the feature ✓). A stronger control would be a second allowed namespace and verifying the non-excluded source does auto-reflect across into it in the same window `DelayForReflection` gives. Optionally a second test case for when the exclusion pattern matches a *target* namespace (source elsewhere) — same watcher-filter path, worth pinning down.

File: src/helm/reflector/values.yaml [hunk 7]
```diff
@@ -28,6 +28,10 @@ configuration:
     minimumLevel: Information
   watcher:
     timeout: ""
+    # Comma-separated list of namespace glob patterns to exclude from watching.
```
Comment (winromulus) [inline src/helm/reflector/values.yaml]: Minor consistency: the README was updated to "exclude from reflection processing" but this helm comment still says "exclude from watching." Next reader going from helm values → README shouldn't see conflicting wording.

**Review by winromulus — DISMISSED** [review]
Really appreciate the engagement here — this is a clean, well-tested response. Walked through each of the earlier asks against the current code:

- **Drop redundant `.Where`** → `GlobMatcher.ParseGlobPatterns` uses only `Split(RemoveEmptyEntries | TrimEntries).Select(...)`. ✓
- **README wording** → now "exclude from reflection processing." ✓
- **Case sensitivity** → patterns normalized to lowercase at parse time. ✓
- **NamespaceWatcher no-op comment** → clear 3-line explanation inline. ✓
- **Extract to helper class** → `GlobMatcher` with a clean static surface. ✓
- **Unit tests** → 19 tests in `GlobMatcherTests` with good coverage; the metacharacter cases (`ns.special` as literal, `ns[1]` with brackets) are the ones that would catch a future refactor breaking the escape-before-substitute order. Nice.
- **Integration test** → `ExcludedNamespacesIntegrationTests` with its own fixture wiring. ✓

On the AI-assisted tests: honest take — the unit tests are genuinely solid, especially the metacharacter edge cases. The integration test is functionally correct with one structural observation left inline.

Three small things, none blocking — all covered inline. Otherwise this is in really good shape.

**Comment by winromulus** [comment]
Heads up — the channel-buffer bump (`256 → 1024`) just landed on `main` as part of [e6ba44f](https://github.com/emberstack/kubernetes-reflector/commit/e6ba44f) (same commit that closed #624 earlier today). That touches the same area of `WatcherBackgroundService.cs` you're editing, which is where the conflict is coming from. On rebase, take the `1024` value from `main` and keep your exclusion-filter additions around it — nothing else should collide.

**Comment by komapa** [comment]
> Really appreciate the engagement here — this is a clean, well-tested response. Walked through each of the earlier asks against the current code:
> 
> * **Drop redundant `.Where`** → `GlobMatcher.ParseGlobPatterns` uses only `Split(RemoveEmptyEntries | TrimEntries).Select(...)`. ✓
> * **README wording** → now "exclude from reflection processing." ✓
> * **Case sensitivity** → patterns normalized to lowercase at parse time. ✓
> * **NamespaceWatcher no-op comment** → clear 3-line explanation inline. ✓
> * **Extract to helper class** → `GlobMatcher` with a clean static surface. ✓
> * **Unit tests** → 19 tests in `GlobMatcherTests` with good coverage; the metacharacter cases (`ns.special` as literal, `ns[1]` with brackets) are the ones that would catch a future refactor breaking the escape-before-substitute order. Nice.
> * **Integration test** → `ExcludedNamespacesIntegrationTests` with its own fixture wiring. ✓
> 
> On the AI-assisted tests: honest take — the unit tests are genuinely solid, especially the metacharacter edge cases. The integration test is functionally correct with one structural observation left inline.
> 
> Three small things, none blocking — all covered inline. Otherwise this is in really good shape.

I believe all the latest requests have been addressed now. 🙇🏼

**Review by winromulus — APPROVED** [review]
Thanks for sticking with this through the back-and-forth — really appreciate the responsiveness on every piece of feedback.

The expanded integration test landed exactly where I was hoping, and I particularly like that the second test reads like inline documentation for the source-vs-target distinction. The unit tests on the regex metacharacter cases are great too — exactly the kind of thing that'll catch a well-meaning future refactor breaking the escape order.

Approving. Thanks again for solving a real operational pain point for folks running this at scale.
